###Link to the Google Colaboratory notebook

https://colab.research.google.com/drive/1ltOVU_d2oPqbHIWwBsu3FI1yqvRxjsl-?usp=sharing

> *Module Title:	Machine Learning*

> *Module Code:	 B9DA109*

> *Module Leader:	Dr Shazia A Afzal*

> *Lecturer:	Mr Kunwar Madan*

> *Assessment Title:	Machine Learning Tasks*

> *Assessment Number (if relevant): 2*

> *Restrictions on Time/Length :	Submission before deadline*

> *Individual/Group: Group (max 3 students)*
* *Individual Tasks To be done Individually*

> *Group members:*	
* *Daniel Oduniyi - 10577159*
* *Stephanie Eje - 10576098*
* *Arnold Chinoso - 10581961*

> *Assessment Weighting: 60%*

> *Issue Date: 3rd December 2021*

> *Hand In Date:	23rd December 2021*

> *Mode of Submission:*
* *Moodle
Task 1 must be submitted on a separate link on Moodle named as Group name – Task 1 Submission
Task 2 must be submitted on the link CA_TWO_(60%)*

##Task 2

###Importing necessary libraries

In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
import plotly.graph_objs as go
import plotly.figure_factory as ff

In [ ]:
# Importing dataset and examining it
df = pd.read_csv("/content/drive/MyDrive/sign_mnist.csv")
df.head()

label  pixel1  pixel2  pixel3  ...  pixel781  pixel782  pixel783  pixel784
0      3     107     118     127  ...       206       204       203       202
1      6     155     157     156  ...       175       103       135       149
2      2     187     188     188  ...       198       195       194       195
3      2     211     211     212  ...       225       222       229       163
4     13     164     167     170  ...       157       163       164       179

[5 rows x 785 columns]

In [ ]:
# Printing the number of row and columns
df.shape

(10000, 785)

In [ ]:
# More information on the dataset and datatypes contained
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Columns: 785 entries, label to pixel784
dtypes: int64(785)
memory usage: 59.9 MB


Checking to see if the dataset is balanced

In [ ]:
labels = list(df.label.unique())
len(labels)

24

In [ ]:
lableCountlst = []
for i in labels:
  dflabelCount = df.loc[(df['label'] == i)]
  x,y = dflabelCount.shape
  lableCountlst.append('{}:{}'.format(i,x))

print('label : number of instances')
print(*lableCountlst, sep = ', ')

label : number of instances
3:442, 6:403, 2:419, 13:411, 16:468, 8:438, 22:425, 18:448, 10:393, 20:428, 17:478, 19:442, 21:405, 23:433, 24:399, 1:377, 12:370, 11:438, 15:383, 4:348, 0:403, 5:424, 7:389, 14:436


The dataset is imbalanced so there is a need to run SMOTE

In [ ]:
# Dividing dataset into label and feature sets
X = df.drop(['label'], axis = 1) # Features
Y = df['label'] # Labels

In [ ]:
print(type(X))
print(type(Y))
print(X.shape)
print(Y.shape)

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>
(10000, 784)
(10000,)


###Dimensionality Reduction Techniques

Ideally, because there are so many variables in the dataset, it would be best to reduce the number of dimensions, to visualize the dataset, prevent overfitting and increase the speed of the model.

There are two ways this can be accomplished.

> Principal component Analysis (PCA)

> Linear Discriminant Analysis (LDA)








In [ ]:
# Normalizing numerical features so that each feature has mean 0 and variance 1
feature_scaler = StandardScaler()
X_scaled = feature_scaler.fit_transform(X)

In [ ]:
# Implementing PCA to visualize dataset
pca = PCA(n_components = 50) 
# PCA can make as many principle components as there are dimensions in the dataset
# That is exactly what would happen if the n_components was not specified
pca.fit(X_scaled)
x_pca = pca.transform(X_scaled)

print("Variance explained by each of the n_components: ",pca.explained_variance_ratio_)
print("Total variance explained by the n_components: ",sum(pca.explained_variance_ratio_))

Variance explained by each of the n_components:  [0.35699428 0.08360827 0.07030617 0.04936101 0.03656472 0.02660766
 0.02543923 0.0205608  0.01821838 0.01626431 0.01377748 0.01276032
 0.01107275 0.01046889 0.00884994 0.0079053  0.00753142 0.00713383
 0.00676892 0.00601923 0.00545875 0.00511563 0.00492648 0.00471673
 0.00432814 0.00417515 0.00390926 0.00367833 0.00343557 0.0033528
 0.00327303 0.00313677 0.00307424 0.00293781 0.00285832 0.00280837
 0.00265904 0.00261485 0.00251574 0.00246615 0.00232606 0.002255
 0.00217776 0.00208058 0.00199635 0.00197095 0.00192958 0.00190261
 0.00181341 0.00177191]
Total variance explained by the n_components:  0.8878783083523706


In [ ]:
# Plotting a scatter graph to see how the first two Principal Components classify the data
digits=list(df['label'])
data = [go.Scatter(x=x_pca[:,0], y=x_pca[:,1], mode='markers',
                    marker = dict(color=Y, colorscale='rainbow', opacity=0.5),
                                text=[f'digit: {a}' for a in digits],
                                hoverinfo='text')]

layout = go.Layout(title = 'PCA Dimensionality Reduction', width = 700, height = 700,
                    xaxis = dict(title='First Principal Component'),
                    yaxis = dict(title='Second Principal Component'))
fig = go.Figure(data=data, layout=layout)
fig.show()

PCA is an unsupervised machine learning algorithm that does not take the target variable into acccount. It only finds the direction in data in which the data varies the most.

As seen in the output, the 50 principal components have captured 88% of the variance in the dataset.

However, when we plot the first two principal components against each other, we can see that because the target labels were not taken into account, they are not doing a good job of actually separating the classes

In [ ]:
# Implementing LDA to visualize dataset
lda = LDA()
# LDA can create as many linear discriminants as there are classes - 1
# max linear discriminants = n_classes - 1
# This is exactly what would take place n_components was not passed into the function
x_lda = lda.fit_transform(X_scaled, Y)

print("Variance explained by each of the n_components: ",lda.explained_variance_ratio_)
print("Total variance explained by the n_components: ",sum(lda.explained_variance_ratio_))

Variance explained by each of the n_components:  [0.13532317 0.12233362 0.09801533 0.07184491 0.06365342 0.05086866
 0.04343683 0.04231177 0.04166021 0.0378902  0.03381986 0.03309612
 0.02861236 0.02603501 0.02404827 0.02242514 0.02210867 0.02046495
 0.01921732 0.01863169 0.01586114 0.01562177 0.01271958]
Total variance explained by the n_components:  1.0


In [ ]:
# Plotting a scatter graph to see how the first two Linear Discriminants classify the data
data = [go.Scatter(x=x_lda[:,0], y=x_lda[:,1], mode='markers',
                    marker = dict(color=Y, colorscale='Rainbow', opacity=0.5),
                                text=[f'digit: {a}' for a in digits],
                                hoverinfo='text')]

layout = go.Layout(title = 'LDA Dimensionality Reduction', width = 700, height = 700,
                    xaxis = dict(title='Discriminant 1'),
                    yaxis = dict(title='Discriminant 2'))
fig = go.Figure(data=data, layout=layout)
fig.show()

LDA however is a supervised machine algorithm used in dimensionality reduction. It creates a new axis the maximises the distance between the means of each class and mininises the variance within each class.

The first output shows that the 23 linear discriminants account for 100% of the variance in the data.

Moreover, when the first two linear discriminants are plotted against each other the graph shows that the data is being very well classified. 

###Classification models

In this section three (3) classifiers will be implemented to model the data

> Support Vector Classifier

> Random Forest Classifier

> Artificial Neural Network: Multi-layer Perceptron (MLP) 

In [ ]:
from sklearn.model_selection import GridSearchCV
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import RandomForestClassifier
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

Implementing Support Vector Classifier

In [ ]:
# We are building a pipeline, which takes in SMOTE (Synthetic Minority Oversampling TEchnique) 
# and the Classification model, SVC (Support Vector Classifier).

# SVC a line to separate the data supported using only a few data points (vectors).
# SMOTE is being implented here because the dataset is imbalanced.

model = Pipeline([
        ('balancing', SMOTE(random_state = 101)),
        ('classification', SVC(random_state=1))
    ])

# Because we do not know the best parameters to use in developing the model, 
# we would run a grid search with a list of parameters and then pick the model
# that gives us the best score
# We are trying all the 4 kernels,
# and, classification__C here refers to regularization term (penalty factor) in SV machines

grid_param = {'classification__kernel': ['linear','poly','rbf','sigmoid'], 'classification__C': [.001,.01,.1,1,10,100]}

# Scoring was set to accuracy because we would want a model that will accurately predict
# the hand gestures correctly

# cv = 5 means we a re using 5 - fold cross validation

gd_sr = GridSearchCV(estimator=model, param_grid=grid_param, scoring='accuracy', cv=5)

# Fitting the model to our data

gd_sr.fit(x_lda, Y)

best_parameters = gd_sr.best_params_ # This shows us the parameters that gives us the best score
print("Best parameters: ", best_parameters)

best_result = gd_sr.best_score_ # Mean cross-validated score of the best_estimator
print("Best result: ", best_result)

Best parameters:  {'classification__C': 10, 'classification__kernel': 'rbf'}
Best result:  0.9991


Implementing Decision Tree Classifier

In [ ]:
# We are building a pipeline, which takes in SMOTE (Synthetic Minority Oversampling TEchnique) 
# and the Classification model, Random Forest Classifier.
# SMOTE is being implented here because the dataset is imbalanced.

# We are using entropy as the crieteria for constructing out tree
# Meaning that the split that gives us the best information gain
# will be used for the split

# max_features has been set to auto, this means thst
# The square root of the total number if features will be used considered when
# looking for the bestsplit

model = Pipeline([
        ('balancing', SMOTE(random_state = 101)),
        ('classification', RandomForestClassifier(criterion='entropy', max_features='auto', random_state=1) )
    ])

# n_estimators here represents the total number of trees that a developed in parallel.
# As a general rule of thumb this number should not exceed 100.
# That being said, although we are using [75,100,125] as our parameter,
# we had run the code before using [10,20,30,40,50,100] and 100 was selected as the best parameter.

grid_param = {'classification__n_estimators': [75,100,125]}

gd_sr = GridSearchCV(estimator=model, param_grid=grid_param, scoring='accuracy', cv=5)

gd_sr.fit(x_lda, Y)

best_parameters = gd_sr.best_params_ # This shows us the parameters that gives us the best score
print(best_parameters)

best_result = gd_sr.best_score_ # Mean cross-validated score of the best_estimator
print(best_result)

{'classification__n_estimators': 100}
0.9983000000000001


Implementing Multi-layer Perceptron (MLP)

In [ ]:
# We are building a pipeline, which takes in SMOTE (Synthetic Minority Oversampling TEchnique) 
# and the Classification model, Multi-layer Perceptron Classifier.
# SMOTE is being implented here because the dataset is imbalanced.

model = Pipeline([
        ('balancing', SMOTE(random_state = 101)),
        ('classification', MLPClassifier(random_state=1)) 
    ])

# We are using four hyper parameters
# classification__activation: the activation functions of the hidden layer(s)
# classification__learning_rate_init: controls the step-size in updating the weights.
# classification__max_iter: the maximum number of iterations
# classification__hidden_layer_sizes: the number of hidden layers and nodes in each layer
# (100,) - one hidden layer with 100 nodes
# (100,100) - two hidden layers with 100 nodes each

grid_param = {'classification__activation': ['identity','logistic', 'tanh', 'relu'],
              'classification__learning_rate_init': [0.0001, 0.001, 0.01, 0.1, 1, 10], 'classification__max_iter' : [750,1000,1250],
              'classification__hidden_layer_sizes': [(100,),(100,100)]}

gd_sr = GridSearchCV(estimator=model, param_grid=grid_param, scoring='accuracy', cv=5)

gd_sr.fit(x_lda, Y)

best_parameters = gd_sr.best_params_ # This shows us the parameters that gives us the best score
print(best_parameters)

best_result = gd_sr.best_score_ # Mean cross-validated score of the best_estimator
print(best_result)

{'classification__activation': 'tanh', 'classification__hidden_layer_sizes': (100,), 'classification__learning_rate_init': 0.01, 'classification__max_iter': 750}
0.9992000000000001


All three of the models are performing excellently, doing a good job of accurately predicting the hand gestures.

The best performing model is the Multi-layer perceptron with an accuracy score of 99.92%

The reason all models are giving good scores could be that the Linear Discriminant accounts for all the variance in the data. Meaning, all the data points have a very clear separation.

###Bibliography

Imbalanced learn, (2021). ‘SMOTE.’ Available at: https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOTE.html#r001eabbe5dd7-1 (Last accessed 16th December 2021)

scikit learn, (2021). ‘sklearn.svm.SVC.’ Available at: https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html?highlight=svc#sklearn.svm.SVC (Last accessed 16th December 2021)

scikit learn, (2021). ‘sklearn.ensemble.RandomForestClassifier.’ Available at: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html?highlight=random%20forest#sklearn.ensemble.RandomForestClassifier (Last accessed 16th December 2021)

scikit learn, (2021). ‘sklearn.neural_network.MLPClassifier.’ Available at: https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html?highlight=mlp#sklearn.neural_network.MLPClassifier (Last accessed 16th December 2021)

